fontforgeexe/shapers/harfbuzz.cpp

```diff
@@ -0,0 +1,349 @@
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "harfbuzz.hpp"
+
+#include <algorithm>
+#include <cassert>
+
+extern "C" {
+#include "gfile.h"
+#include "splinechar.h"
+#include "splinefont_enums.h"
+#include "utype.h"
+}
+
+namespace ff::shapers {
+
+HarfBuzzShaper::HarfBuzzShaper(std::shared_ptr<ShaperContext> context)
+    : context_(context) {
+    FILE* ttf_file = GFileTmpfile();
+
+    _WriteTTFFont(
+        ttf_file, context_->sf, ff_ttf, NULL, bf_ttf,
+        ttf_flag_otmode | ttf_flag_oldkernmappedonly | ttf_flag_fake_map,
+        context_->get_enc_map(context_->sf), ly_fore);
+
+    // Calculate file length
+    fseek(ttf_file, 0L, SEEK_END);
+    long bufsize = ftell(ttf_file);
+    fseek(ttf_file, 0L, SEEK_SET);
+
+    // Read the entire file into memory
+    blob = (char*)malloc(sizeof(char) * (bufsize + 1));
+    size_t blob_size = fread(blob, sizeof(char), bufsize, ttf_file);
+
+    hb_ttf_blob =
+        hb_blob_create(blob, blob_size, HB_MEMORY_MODE_WRITABLE, NULL, NULL);
+
+    hb_ttf_face = hb_face_create(hb_ttf_blob, 0);
+
+    hb_ttf_font = hb_font_create(hb_ttf_face);
+
+    fclose(ttf_file);
+}
+
+HarfBuzzShaper::~HarfBuzzShaper() {
+    hb_font_destroy(hb_ttf_font);
+    hb_face_destroy(hb_ttf_face);
+    hb_blob_destroy(hb_ttf_blob);
+    free(blob);
+}
+
+SplineChar** HarfBuzzShaper::extract_shaped_data(hb_buffer_t* hb_buffer) {
+    unsigned int glyph_count;
+    hb_glyph_info_t* glyph_info_arr =
+        hb_buffer_get_glyph_infos(hb_buffer, &glyph_count);
+    hb_glyph_position_t* glyph_pos_arr =
+        hb_buffer_get_glyph_positions(hb_buffer, &glyph_count);
+
+    // Return NULL-terminated raw C-style array of pointers
+    SplineChar** glyphs_after_gpos =
+        (SplineChar**)calloc(glyph_count + 1, sizeof(SplineChar*));
+
+    // Adjust metrics buffer size
+    metrics.resize(glyph_count + 1);
+
+    // Process the glyphs and positions
+    int total_x_advance = 0, total_y_advance = 0;
+    for (int i = 0; i < glyph_count; ++i) {
+        char glyph_name[64];
+        hb_glyph_info_t& glyph_info = glyph_info_arr[i];
+        hb_glyph_position_t& glyph_pos = glyph_pos_arr[i];
+
+        // Warning: after the shaping glyph_info->codepoint is not a Unicode
+        // point, but rather an internal glyph index. We can't use it in our
+        // functions.
+        hb_bool_t found =
+            hb_font_get_glyph_name(hb_ttf_font, glyph_info.codepoint,
+                                   glyph_name, sizeof(glyph_name) - 1);
+        SplineChar* glyph_out =
+            context_->get_glyph_by_name(context_->sf, -1, glyph_name);
+
+        glyphs_after_gpos[i] = glyph_out;
+
+        // Fill unscaled metrics in font units
+        hb_position_t h_advance =
+            hb_font_get_glyph_h_advance(hb_ttf_font, glyph_info.codepoint);
+
+        hb_glyph_extents_t extents;
+        hb_bool_t res = hb_font_get_glyph_extents(
+            hb_ttf_font, glyph_info.codepoint, &extents);
+        assert(res);
```

Comment:
Calculating glyph extents can be slow (especially for CFF fonts), and I don’t see extents used anywhere so it seems wasteful.